implement significand in native Julia

simonbyrne opened the pull request:
`significand` utilised non-standard libm functions, which will be removed from the next release of Openlibm (https://github.com/JuliaLang/openlibm/pull/85). This implements them in native Julia.

(I also changed some comparisons from `x==zero(x)` to `x==0`: LLVM now optimises these away (see #9030); this change is purely for notational convenience)

simonbyrne committed:
implement significand, exponent and frexp in native Julia: use consistent functions across all types.

kmsquire commented:
Any reason not to also add a version for Float16?

simonbyrne commented:
@kmsquire No, only because we didn't have them before. Now done.

quinnj commented:
This is great to have for the grisu code as well. We currently promote Float16s up because there wasn't a significand defined. I'll have to dig back into that code soon and do another pass.

jiahao commented:
bump

+1

simonbyrne commented:
Updated.

ViralBShah commented:
Should we be adding more tests to ensure full coverage - especially since we have our own implementation (even though it is pretty simple)?

simonbyrne commented:
Probably, I'll see what I can do.

simonbyrne commented:
I got a bit carried away, and implemented general methods for `significand`, `exponent` and `frexp` across the different types. I also added some more tests.

StefanKarpinski commented:
wow! very cool. feel free to merge whenever you feel confident that this is correct and passing tests.

quinnj commented:
Awesome work as usual @simonbyrne. With all three defined for our float types, I can probably cut about 250 lines out of the grisu code. I'll have to do another pass to work that out once this gets merged.

simonbyrne commented:
I can't get the osx test to pass, but this doesn't seem to be related.
